File: xnote.js
// ── State ──────────────────────────────────────────────────────────────────
let currentNoteId = null;
let activeTab = 'recent';
let allNotesPage = 1;

// ── View management ────────────────────────────────────────────────────────
function showView(id) {
    document.querySelectorAll('.view').forEach(v => v.classList.remove('active'));
    document.getElementById(id).classList.add('active');
}

async function showHome() {
    currentNoteId = null;
    showView('homeView');
    document.getElementById('searchInput').value = '';
    closeDropdown();
    if (activeTab === 'recent') {
        await loadRecentNotes();
    } else {
        await loadAllNotes();
    }
    updateTabUI();
}

function switchTab(tab) {
    activeTab = tab;
    allNotesPage = 1;
    updateTabUI();
    if (tab === 'recent') {
        loadRecentNotes();
    } else {
        loadAllNotes();
    }
}

function updateTabUI() {
    document.getElementById('tabRecent').classList.toggle('active', activeTab === 'recent');
    document.getElementById('tabAll').classList.toggle('active', activeTab === 'all');
}

async function loadRecentNotes() {
    const notes = await app.call('get_recent_notes');
    renderNoteList(notes);
    document.getElementById('pagination').classList.remove('visible');
}

async function loadAllNotes(page) {
    if (page !== undefined) allNotesPage = page;
    const data = await app.call('get_all_notes', {page: allNotesPage});
    renderNoteList(data.notes);
    renderPagination(data);
}

function renderNoteList(notes) {
    const el = document.getElementById('noteList');
    if (!notes || notes.length === 0) {
        el.innerHTML = `
            <div class="empty-state">
                <div class="big">⬡</div>
                No notes yet — press <strong>+ New</strong> to create your first one.
            </div>`;
        return;
    }
    el.innerHTML = notes.map(n => `
        <div class="note-row" onclick="openNote(${n.id})">
            <span class="title">${esc(n.title)}</span>
            <span class="date">${n.date_viewed.slice(0, 10)}</span>
        </div>
    `).join('');
}

function renderPagination(data) {
    const el = document.getElementById('pagination');
    if (data.total_pages <= 1) {
        el.classList.remove('visible');
        return;
    }
    const prevDisabled = data.page <= 1 ? 'disabled' : '';
    const nextDisabled = data.page >= data.total_pages ? 'disabled' : '';
    el.innerHTML = `
        <button class="btn btn-ghost" ${prevDisabled} onclick="loadAllNotes(${data.page - 1})">← Prev</button>
        <span class="page-info">${data.page} / ${data.total_pages}</span>
        <button class="btn btn-ghost" ${nextDisabled} onclick="loadAllNotes(${data.page + 1})">Next →</button>
    `;
    el.classList.add('visible');
}

// ── Search ─────────────────────────────────────────────────────────────────
let searchTimer = null;
function onSearch(q) {
    clearTimeout(searchTimer);
    if (!q.trim()) { closeDropdown(); return; }
    searchTimer = setTimeout(async () => {
        const results = await app.call('search_notes', {query: q});
        const dd = document.getElementById('searchDropdown');
        if (!results || results.length === 0) { closeDropdown(); return; }
        dd.innerHTML = results.map(r => `
            <div class="dropdown-item" onclick="openNote(${r.id})">
                <span class="title">${esc(r.title)}</span>
                <span class="date">${r.date_modified.slice(0, 10)}</span>
            </div>
        `).join('');
        dd.classList.add('open');
    }, 150);
}
function onSearchFocus() {
    const q = document.getElementById('searchInput').value;
    if (q.trim()) onSearch(q);
}
function closeDropdown() {
    document.getElementById('searchDropdown').classList.remove('open');
}

// Close dropdown on outside click
document.addEventListener('click', e => {
    if (!e.target.closest('.search-wrap')) closeDropdown();
});

// ── Editor ─────────────────────────────────────────────────────────────────
async function showEditor(noteId) {
    currentNoteId = noteId || null;
    document.getElementById('editorTitle').value = '';
    document.getElementById('editorTags').value = '';
    document.getElementById('editorBody').value = '';

    if (noteId) {
        const note = await app.call('get_note', {note_id: noteId});
        if (note) {
            document.getElementById('editorTitle').value = note.title;
            document.getElementById('editorTags').value = (note.tags || []).join(', ');
            document.getElementById('editorBody').value = note.body || '';
        }
    }
    showView('editorView');
    document.getElementById('editorTitle').focus();
}

async function saveNote() {
    const title = document.getElementById('editorTitle').value.trim();
    if (!title) {
        document.getElementById('editorTitle').focus();
        document.getElementById('editorTitle').style.borderColor = 'var(--red)';
        setTimeout(() => document.getElementById('editorTitle').style.borderColor = '', 1500);
        return;
    }
    const body = document.getElementById('editorBody').value;
    const tags = document.getElementById('editorTags').value;
    const savedId = await app.call('save_note', {title: title, body: body, tags_str: tags, note_id: currentNoteId});
    openNote(savedId);
}

async function attachImage() {
    const tag = await app.pickAndUpload('image/*');
    if (tag) {
        const ta = document.getElementById('editorBody');
        const start = ta.selectionStart;
        const before = ta.value.slice(0, start);
        const after = ta.value.slice(ta.selectionEnd);
        ta.value = before + tag + after;
        ta.selectionStart = ta.selectionEnd = start + tag.length;
        ta.focus();
    }
}

// ── Viewer ─────────────────────────────────────────────────────────────────
async function openNote(noteId) {
    closeDropdown();
    const note = await app.call('get_note', {note_id: noteId});
    if (!note) return;
    currentNoteId = noteId;

    document.getElementById('viewerTitle').textContent = note.title;

    // Meta chips
    let meta = `<span class="chip">created ${note.date_created.slice(0,10)}</span>`;
    meta += `<span class="chip">modified ${note.date_modified.slice(0,10)}</span>`;
    meta += `<span class="chip">viewed ${note.date_viewed.slice(0,10)}</span>`;
    (note.tags || []).forEach(t => { meta += `<span class="chip tag"># ${esc(t)}</span>`; });
    document.getElementById('viewerMeta').innerHTML = meta;

    document.getElementById('viewerBody').innerHTML = prepareBody(note.body || '');

    showView('viewerView');
}

function editCurrent() {
    if (currentNoteId) showEditor(currentNoteId);
}

async function popOutNote() {
    if (!currentNoteId) return;
    window.open('/page/note?note_id=' + currentNoteId, '_blank');
}

// ── Delete ─────────────────────────────────────────────────────────────────
function confirmDelete() {
    document.getElementById('deleteOverlay').classList.add('open');
}
function closeDelete() {
    document.getElementById('deleteOverlay').classList.remove('open');
}
async function doDelete() {
    if (currentNoteId) {
        await app.call('delete_note', {note_id: currentNoteId});
        closeDelete();
        showHome();
    }
}

// ── Helpers ────────────────────────────────────────────────────────────────
function esc(s) {
    const d = document.createElement('div');
    d.textContent = s;
    return d.innerHTML;
}

const HTML_TAG_RE = /<\/?(a|abbr|article|aside|b|blockquote|br|code|col|dd|del|details|div|dl|dt|em|figcaption|figure|footer|form|h[1-6]|header|hr|i|img|input|ins|li|link|main|mark|nav|ol|p|pre|q|s|section|small|span|strong|style|sub|summary|sup|table|tbody|td|textarea|tfoot|th|thead|tr|u|ul|var|video)[\/\s>]/i;

function isHtml(text) {
    return HTML_TAG_RE.test(text);
}

function prepareBody(text) {
    if (isHtml(text)) return text;
    // Plain text: escape and convert newlines
    return text
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/\n/g, '<br>\n');
}

// ── Config ─────────────────────────────────────────────────────────────────
const PRESETS = {
    'xnote Default': {
        bg: '#0a0a0f', bg2: '#16161e', bg3: '#1e1e2a', entry: '#111118',
        border: '#2a2a3a', fg: '#e8e8f0', fg2: '#6a6a80', accent: '#77ccff',
        accent2: '#b07aff', green: '#3ad900', red: '#ff3366', yellow: '#fbbf24', code: '#ff0099'
    },
    'Cobalt': {
        bg: '#001b33', bg2: '#003b70', bg3: '#0065bf', entry: '#00213f',
        border: '#003b70', fg: '#ffffff', fg2: '#cccccc', accent: '#77ccff',
        accent2: '#80ffbb', green: '#3ad900', red: '#ff0044', yellow: '#ffee80', code: '#ff0099'
    },
    'Warm Noir': {
        bg: '#111010', bg2: '#1c1917', bg3: '#292524', entry: '#161412',
        border: '#3d3530', fg: '#f5f0eb', fg2: '#8a7f72', accent: '#e8a04c',
        accent2: '#c47adb', green: '#4ade80', red: '#ef4444', yellow: '#fcd34d', code: '#fb7185'
    },
    'Forest': {
        bg: '#0a120e', bg2: '#121f18', bg3: '#1a2e24', entry: '#0e1812',
        border: '#2a3f32', fg: '#e0efe6', fg2: '#6b8a76', accent: '#5ce0a0',
        accent2: '#a0d4f7', green: '#4ade80', red: '#f87171', yellow: '#fde68a', code: '#f472b6'
    }
};

const COLOR_LABELS = {
    bg: 'Background (root)', bg2: 'Background (panels)', bg3: 'Background (hover)',
    entry: 'Input surfaces', border: 'Borders & separators', fg: 'Primary text',
    fg2: 'Muted text', accent: 'Accent (logo, buttons, headings)',
    accent2: 'Accent 2 (blockquote, tags)', green: 'Edit / success',
    red: 'Delete / danger', yellow: 'Strong text / warnings', code: 'Inline code text'
};

const FONTS = [
    'IBM Plex Mono', 'Fira Code', 'Cascadia Code', 'JetBrains Mono',
    'Ubuntu Mono', 'DejaVu Sans Mono', 'Liberation Mono', 'Noto Mono',
    'Hack', 'Fira Mono', 'Source Code Pro', 'Courier New', 'monospace'
];

let configWorking = {};
let configFont = '';
let configRecentLimit = 5;
let configAllNotesPageSize = 10;
let configAllNotesSort = 'date_viewed';
let configDirty = false;

async function showConfig() {
    // Snapshot current CSS vars as working copy
    const style = getComputedStyle(document.documentElement);
    configWorking = {};
    for (const role of Object.keys(COLOR_LABELS)) {
        const val = style.getPropertyValue('--' + role).trim();
        configWorking[role] = val || PRESETS['xnote Default'][role];
    }
    configFont = style.getPropertyValue('--font').split(',')[0].replace(/'/g, '').trim();
    configDirty = false;

    // Load recent limit
    const limit = await app.call('get_recent_limit');
    configRecentLimit = limit;
    document.getElementById('recentLimitSelect').value = String(limit);

    // Load all notes settings
    const allPageSize = await app.call('get_all_notes_page_size');
    configAllNotesPageSize = allPageSize;
    document.getElementById('allNotesPageSizeSelect').value = String(allPageSize);

    const allSort = await app.call('get_all_notes_sort');
    configAllNotesSort = allSort;
    document.getElementById('allNotesSortSelect').value = allSort;

    // Build preset buttons
    const presetBar = document.getElementById('presetBar');
    presetBar.innerHTML = Object.keys(PRESETS).map(name =>
        `<button class="preset-btn" onclick="loadPreset('${name}')">${name}</button>`
    ).join('');

    // Build font select
    const fontSelect = document.getElementById('fontSelect');
    fontSelect.innerHTML = FONTS.map(f =>
        `<option value="${f}" ${f === configFont ? 'selected' : ''}>${f}</option>`
    ).join('');
    document.getElementById('fontPreview').style.fontFamily = configFont;

    // Build color rows
    buildColorRows();

    showView('configView');
}

function buildColorRows() {
    const container = document.getElementById('colorRows');
    container.innerHTML = Object.entries(COLOR_LABELS).map(([role, label]) => `
        <div class="config-row">
            <label>${label}</label>
            <div class="color-swatch" id="swatch-${role}"
                 style="background:${configWorking[role]}"
                 onclick="pickColor('${role}')"></div>
            <span class="color-hex" id="hex-${role}">${configWorking[role]}</span>
            <input type="color" class="color-input-hidden" id="picker-${role}"
                   value="${configWorking[role]}" onchange="onColorPicked('${role}', this.value)" />
        </div>
    `).join('');
}

function pickColor(role) {
    document.getElementById('picker-' + role).click();
}

function onColorPicked(role, value) {
    configWorking[role] = value;
    document.getElementById('swatch-' + role).style.background = value;
    document.getElementById('hex-' + role).textContent = value;
    configDirty = true;
}

function onFontChange(family) {
    configFont = family;
    document.getElementById('fontPreview').style.fontFamily = family;
    configDirty = true;
}

function onRecentLimitChange(value) {
    configRecentLimit = parseInt(value);
    configDirty = true;
}

function onAllNotesPageSizeChange(value) {
    configAllNotesPageSize = parseInt(value);
    configDirty = true;
}

function onAllNotesSortChange(value) {
    configAllNotesSort = value;
    configDirty = true;
}

function loadPreset(name) {
    const preset = PRESETS[name];
    if (!preset) return;
    configWorking = {...preset};
    buildColorRows();
    // Highlight active preset
    document.querySelectorAll('.preset-btn').forEach(b => {
        b.classList.toggle('active', b.textContent === name);
    });
    configDirty = true;
}

function resetConfig() {
    loadPreset('xnote Default');
    document.getElementById('fontSelect').value = 'IBM Plex Mono';
    configFont = 'IBM Plex Mono';
    document.getElementById('fontPreview').style.fontFamily = configFont;
    configRecentLimit = 5;
    document.getElementById('recentLimitSelect').value = '5';
    configAllNotesPageSize = 10;
    document.getElementById('allNotesPageSizeSelect').value = '10';
    configAllNotesSort = 'date_viewed';
    document.getElementById('allNotesSortSelect').value = 'date_viewed';
}

async function applyConfig() {
    // Apply CSS variables live
    const root = document.documentElement;
    for (const [role, val] of Object.entries(configWorking)) {
        root.style.setProperty('--' + role, val);
    }
    root.style.setProperty('--font', `'${configFont}', monospace`);

    // Persist to DB
    await app.call('save_palette', {palette: configWorking});
    await app.call('save_font_family', {family: configFont});
    await app.call('save_recent_limit', {limit: configRecentLimit});
    await app.call('save_all_notes_page_size', {size: configAllNotesPageSize});
    await app.call('save_all_notes_sort', {sort_by: configAllNotesSort});

    showHome();
}

function cancelConfig() {
    showHome();
}

// ── Right-click to copy (desktop only — mobile uses native selection) ──────
function copyToClipboard(text) {
    const ta = document.createElement('textarea');
    ta.value = text;
    ta.style.position = 'fixed';
    ta.style.opacity = '0';
    document.body.appendChild(ta);
    ta.select();
    document.execCommand('copy');
    document.body.removeChild(ta);
}

document.addEventListener('contextmenu', e => {
    // On touch devices, let the native selection toolbar handle copy
    if ('ontouchstart' in window) return;
    if (!e.target.closest('.viewer-body-inner')) return;
    const sel = window.getSelection().toString();
    if (!sel) return;
    e.preventDefault();
    copyToClipboard(sel);
});

// ── Init ───────────────────────────────────────────────────────────────────
document.addEventListener('DOMContentLoaded', async () => {
    // Load saved palette
    const palette = await app.call('get_palette');
    if (palette) {
        const root = document.documentElement;
        for (const [role, val] of Object.entries(palette)) {
            root.style.setProperty('--' + role, val);
        }
    }
    // Load saved font
    const font = await app.call('get_font_family');
    if (font) {
        document.documentElement.style.setProperty('--font', `'${font}', monospace`);
    }
    showHome();
});

// Keyboard shortcuts
document.addEventListener('keydown', e => {
    if (e.ctrlKey && e.key === 'n') { e.preventDefault(); showEditor(); }
    if (e.key === 'Escape') {
        if (document.getElementById('deleteOverlay').classList.contains('open')) {
            closeDelete();
        } else if (document.getElementById('configView').classList.contains('active')) {
            cancelConfig();
        } else if (document.getElementById('editorView').classList.contains('active')) {
            showHome();
        } else if (document.getElementById('viewerView').classList.contains('active')) {
            showHome();
        }
    }
});

File: xnote.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.11"
# dependencies = [
#     "flask",
#     "pywebview",
#     "qtpy",
#     "PyQt6",
#     "PyQt6-WebEngine",
# ]
# ///

"""
xnote — Personal note-taking app (Hearth app).

Stores notes in SQLite database (xnote.db).

Run:
    [uv run] xnote.py                   # native window
    [uv run] xnote.py --serve [PORT]    # LAN web access

Developer:  KarmaHelen
Contact:    xnote.cusp743@passinbox.com
Support:    https://buymeacoffee.com/karmahelen
"""

import datetime
import json
import os
import shutil
import sqlite3
import sys
from contextlib import contextmanager
from pathlib import Path

BASE_DIR = Path(__file__).resolve().parent
DB_PATH = BASE_DIR / "xnote.db"
IMG_DIR = BASE_DIR / "images"

# Add parent directory to path so Hearth can be imported
sys.path.insert(0, str(BASE_DIR.parent))


class XNoteApp:

    def __init__(self):
        IMG_DIR.mkdir(exist_ok=True)
        self._init_db()

    # ── Private helpers ──────────────────────────────────────────────────

    def _init_db(self):
        conn = sqlite3.connect(DB_PATH)
        conn.executescript("""
            PRAGMA foreign_keys = ON;

            CREATE TABLE IF NOT EXISTS notes (
                id            INTEGER PRIMARY KEY AUTOINCREMENT,
                title         TEXT    NOT NULL,
                body          TEXT    DEFAULT '',
                date_created  TEXT    NOT NULL,
                date_modified TEXT    NOT NULL
            );

            CREATE TABLE IF NOT EXISTS tags (
                id   INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT UNIQUE NOT NULL
            );

            CREATE TABLE IF NOT EXISTS note_tags (
                note_id INTEGER NOT NULL REFERENCES notes(id) ON DELETE CASCADE,
                tag_id  INTEGER NOT NULL REFERENCES tags(id)  ON DELETE CASCADE,
                PRIMARY KEY (note_id, tag_id)
            );

            CREATE TABLE IF NOT EXISTS settings (
                key   TEXT PRIMARY KEY,
                value TEXT NOT NULL
            );
        """)

        # Migration: add date_viewed column if it doesn't exist
        cols = [r[1] for r in conn.execute("PRAGMA table_info(notes)").fetchall()]
        if "date_viewed" not in cols:
            conn.execute("ALTER TABLE notes ADD COLUMN date_viewed TEXT")
            conn.execute("UPDATE notes SET date_viewed = date_modified WHERE date_viewed IS NULL")
            conn.commit()

        conn.close()

    @contextmanager
    def _db(self):
        conn = sqlite3.connect(DB_PATH)
        conn.row_factory = sqlite3.Row
        conn.execute("PRAGMA foreign_keys = ON")
        try:
            yield conn
            conn.commit()
        except Exception:
            conn.rollback()
            raise
        finally:
            conn.close()

    def _load_setting(self, key, default=""):
        with self._db() as conn:
            row = conn.execute(
                "SELECT value FROM settings WHERE key=?", (key,)
            ).fetchone()
        return row[0] if row else default

    def _save_setting(self, key, value):
        with self._db() as conn:
            conn.execute(
                "INSERT OR REPLACE INTO settings (key, value) VALUES (?, ?)",
                (key, value)
            )

    def _shutdown(self):
        pass

    # ── Notes CRUD ───────────────────────────────────────────────────────

    def get_recent_notes(self, limit=None):
        if limit is None:
            limit = int(self._load_setting("recent_notes_limit", "5"))
        with self._db() as conn:
            rows = conn.execute(
                "SELECT id, title, date_viewed FROM notes ORDER BY date_viewed DESC LIMIT ?",
                (limit,)
            ).fetchall()
        return [dict(r) for r in rows]

    def get_all_notes(self, page=1, page_size=None, sort_by=None):
        if page_size is None:
            page_size = int(self._load_setting("all_notes_page_size", "10"))
        if sort_by is None:
            sort_by = self._load_setting("all_notes_sort", "date_viewed")

        sort_map = {
            "date_viewed": "date_viewed DESC",
            "date_modified": "date_modified DESC",
            "date_created": "date_created DESC",
            "title": "title COLLATE NOCASE ASC",
        }
        order = sort_map.get(sort_by, "date_viewed DESC")
        offset = (max(1, int(page)) - 1) * page_size

        with self._db() as conn:
            total = conn.execute("SELECT COUNT(*) FROM notes").fetchone()[0]
            rows = conn.execute(
                f"SELECT id, title, date_viewed, date_modified, date_created FROM notes ORDER BY {order} LIMIT ? OFFSET ?",
                (page_size, offset)
            ).fetchall()

        return {
            "notes": [dict(r) for r in rows],
            "total": total,
            "page": max(1, int(page)),
            "page_size": page_size,
            "total_pages": max(1, (total + page_size - 1) // page_size),
        }

    def search_notes(self, query):
        query = query.strip()
        if not query:
            return []

        with self._db() as conn:
            # Priority 1: exact phrase match in title, body, or tags
            exact = f"%{query}%"
            exact_rows = conn.execute("""
                SELECT DISTINCT n.id, n.title, n.date_modified
                FROM notes n
                LEFT JOIN note_tags nt ON n.id = nt.note_id
                LEFT JOIN tags t ON nt.tag_id = t.id
                WHERE n.title LIKE ? OR n.body LIKE ? OR t.name LIKE ?
                ORDER BY n.date_modified DESC
                LIMIT 20
            """, (exact, exact, exact)).fetchall()

            exact_ids = {r['id'] for r in exact_rows}
            results = [dict(r) for r in exact_rows]

            # Priority 2: all individual words must each appear somewhere
            words = query.split()
            if len(words) > 1:
                word_conditions = []
                params = []
                for word in words:
                    w = f"%{word}%"
                    word_conditions.append("""
                        n.id IN (
                            SELECT n2.id FROM notes n2
                            LEFT JOIN note_tags nt2 ON n2.id = nt2.note_id
                            LEFT JOIN tags t2 ON nt2.tag_id = t2.id
                            WHERE n2.title LIKE ? OR n2.body LIKE ? OR t2.name LIKE ?
                        )
                    """)
                    params.extend([w, w, w])

                where = " AND ".join(word_conditions)
                word_rows = conn.execute(f"""
                    SELECT DISTINCT n.id, n.title, n.date_modified
                    FROM notes n
                    WHERE {where}
                    ORDER BY n.date_modified DESC
                    LIMIT 20
                """, params).fetchall()

                for r in word_rows:
                    if r['id'] not in exact_ids:
                        results.append(dict(r))

            return results[:20]

    def get_note(self, note_id):
        now = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        with self._db() as conn:
            conn.execute(
                "UPDATE notes SET date_viewed = ? WHERE id = ?",
                (now, note_id)
            )
            note = conn.execute(
                "SELECT id, title, body, date_created, date_modified, date_viewed FROM notes WHERE id=?",
                (note_id,)
            ).fetchone()
            tags = conn.execute("""
                SELECT t.name FROM tags t
                JOIN note_tags nt ON t.id = nt.tag_id
                WHERE nt.note_id = ?
                ORDER BY t.name
            """, (note_id,)).fetchall()
        if not note:
            return None
        result = dict(note)
        result["tags"] = [r[0] for r in tags]
        return result

    def save_note(self, title, body, tags_str, note_id=None):
        now = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        tag_list = [t.strip().lower() for t in tags_str.split(",") if t.strip()]

        with self._db() as conn:
            if note_id:
                conn.execute(
                    "UPDATE notes SET title=?, body=?, date_modified=? WHERE id=?",
                    (title, body, now, note_id)
                )
            else:
                cur = conn.execute(
                    "INSERT INTO notes (title, body, date_created, date_modified, date_viewed) VALUES (?,?,?,?,?)",
                    (title, body, now, now, now)
                )
                note_id = cur.lastrowid

            conn.execute("DELETE FROM note_tags WHERE note_id=?", (note_id,))
            for tag in tag_list:
                conn.execute("INSERT OR IGNORE INTO tags (name) VALUES (?)", (tag,))
                tag_id = conn.execute("SELECT id FROM tags WHERE name=?", (tag,)).fetchone()[0]
                conn.execute(
                    "INSERT OR IGNORE INTO note_tags (note_id, tag_id) VALUES (?,?)",
                    (note_id, tag_id)
                )
        return note_id

    def delete_note(self, note_id):
        with self._db() as conn:
            conn.execute("DELETE FROM notes WHERE id=?", (note_id,))
        return True

    # ── Page routes (served as HTML via GET /page/note?note_id=...) ──────

    def page_note(self, note_id):
        """Render a note as a standalone HTML page for pop-out viewing."""
        import html as htmlmod
        import re

        note = self.get_note(int(note_id))
        if not note:
            return None

        safe_title = htmlmod.escape(note['title'])
        body = note.get("body", "") or ""

        # Detect if body is HTML or plain text
        html_tags = {'a','b','blockquote','br','code','div','em','h1','h2','h3',
                     'h4','h5','h6','hr','i','img','li','ol','p','pre','span',
                     'strong','sub','sup','table','tbody','td','th','thead','tr','u','ul'}
        is_html = False
        for m in re.finditer(r'</?([a-zA-Z][a-zA-Z0-9]*)', body):
            if m.group(1).lower() in html_tags:
                is_html = True
                break
        if not is_html:
            body = (body
                    .replace("&", "&amp;")
                    .replace("<", "&lt;")
                    .replace(">", "&gt;")
                    .replace("\n", "<br>\n"))

        # Rewrite relative image paths to absolute so they resolve from /page/...
        body = re.sub(r'src="images/', 'src="/images/', body)
        body = re.sub(r"src='images/", "src='/images/", body)

        palette = self.get_palette() or {}
        defaults = {
            'bg': '#0a0a0f', 'bg2': '#16161e', 'bg3': '#1e1e2a', 'entry': '#111118',
            'border': '#2a2a3a', 'fg': '#e8e8f0', 'fg2': '#6a6a80', 'accent': '#77ccff',
            'accent2': '#b07aff', 'green': '#3ad900', 'red': '#ff3366',
            'yellow': '#fbbf24', 'code': '#ff0099'
        }
        for k, v in defaults.items():
            palette.setdefault(k, v)

        font = self.get_font_family()

        return f"""<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>{safe_title}</title>
<style>
*, *::before, *::after {{ box-sizing: border-box; margin: 0; padding: 0; }}
::-webkit-scrollbar {{ width: 6px; }}
::-webkit-scrollbar-track {{ background: transparent; }}
::-webkit-scrollbar-thumb {{ background: {palette['border']}; border-radius: 3px; }}
body {{
    background: {palette['bg']};
    color: {palette['fg']};
    font-family: '{font}', monospace;
    font-size: 14px;
    line-height: 1.8;
    padding: 28px 32px;
    overflow-y: auto;
    height: 100vh;
    user-select: text;
    -webkit-user-select: text;
    cursor: text;
}}
.title {{
    font-size: 20px;
    font-weight: 700;
    color: {palette['fg']};
    margin-bottom: 16px;
    padding-bottom: 12px;
    border-bottom: 1px solid {palette['border']};
}}
h1, h2, h3, h4 {{ color: {palette['accent']}; margin-top: 1.2em; margin-bottom: 0.4em; }}
a {{ color: {palette['accent']}; }}
strong {{ color: {palette['yellow']}; }}
em {{ color: {palette['accent2']}; }}
code {{
    background: {palette['bg3']};
    color: {palette['code']};
    padding: 2px 6px;
    border-radius: 3px;
    font-size: 0.9em;
}}
pre {{
    background: {palette['bg3']};
    border-left: 3px solid {palette['accent']};
    padding: 14px 18px;
    overflow-x: auto;
    border-radius: 4px;
    margin: 12px 0;
}}
pre code {{ background: none; padding: 0; }}
table {{ border-collapse: collapse; width: 100%; margin: 16px 0; }}
th {{ background: {palette['bg3']}; color: {palette['accent']}; padding: 8px 14px; text-align: left; border: 1px solid {palette['border']}; }}
td {{ padding: 8px 14px; border: 1px solid {palette['border']}; }}
tr:nth-child(even) {{ background: {palette['bg2']}; }}
blockquote {{ border-left: 3px solid {palette['accent2']}; margin: 12px 0; padding-left: 18px; color: {palette['fg2']}; }}
img {{ max-width: 100%; border-radius: 4px; margin: 8px 0; }}
ul, ol {{ padding-left: 24px; }}
li {{ margin: 4px 0; }}
hr {{ border: none; border-top: 1px solid {palette['border']}; margin: 20px 0; }}
p {{ margin: 8px 0; }}
</style></head>
<body>
<div class="title">{safe_title}</div>
{body}
</body></html>"""

    # ── File upload (image attachment) ───────────────────────────────────

    def handle_upload(self, filename, filepath):
        """Save uploaded image to the images directory."""
        dest = IMG_DIR / filename
        n = 1
        while dest.exists():
            stem, ext = os.path.splitext(filename)
            dest = IMG_DIR / f"{stem}_{n}{ext}"
            n += 1

        shutil.copy2(filepath, dest)
        return f'<img src="images/{dest.name}" alt="{dest.name}" style="max-width:100%;">'

    # ── Settings ─────────────────────────────────────────────────────────

    def get_palette(self):
        raw = self._load_setting("palette", "")
        if raw:
            try:
                return json.loads(raw)
            except Exception:
                pass
        return None

    def save_palette(self, palette):
        self._save_setting("palette", json.dumps(palette))
        return True

    def get_font_family(self):
        return self._load_setting("font_family", "IBM Plex Mono")

    def save_font_family(self, family):
        self._save_setting("font_family", family)
        return True

    def get_recent_limit(self):
        return int(self._load_setting("recent_notes_limit", "5"))

    def save_recent_limit(self, limit):
        self._save_setting("recent_notes_limit", str(max(1, min(50, int(limit)))))
        return True

    def get_all_notes_page_size(self):
        return int(self._load_setting("all_notes_page_size", "10"))

    def save_all_notes_page_size(self, size):
        self._save_setting("all_notes_page_size", str(max(1, min(50, int(size)))))
        return True

    def get_all_notes_sort(self):
        return self._load_setting("all_notes_sort", "date_viewed")

    def save_all_notes_sort(self, sort_by):
        valid = {"date_viewed", "date_modified", "date_created", "title"}
        if sort_by not in valid:
            sort_by = "date_viewed"
        self._save_setting("all_notes_sort", sort_by)
        return True


# ─────────────────────────────────────────────────────────────────────────────
# Entry Point
# ─────────────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    from hearth import run
    run(XNoteApp(), frontend=str(BASE_DIR / "xnote.html"), title="xnote", port=8080,
        window={
            'width': 740,
            'height': 420,
            'min_size': (500, 300),
            'background_color': '#0a0a0f',
            'text_select': True,
        })
